[SymbolGraph][CursorInfo] Add option to SourceKit's CursorInfo request to include the SymbolGraph JSON

nathawes opened the pull request:
Adds a new `key.retrieve_symbol_graph` option to the request. When set to 1 the JSON for a SymbolGraph containing a single node corresponding to the symbol at the cursor position will be included in the response under `key.symbol_graph`.
    
This extends the SymbolGraph with a new entry point to get a graph for a single symbol, and to additionally support type substitution to match the existing CursorInfo behavior (e.g. so that when invoked on `first` in `Array<Int>().first`, the type is reported as `Int?` rather than `Element?`).
    
Resolves rdar://problem/70551509

Commit:
[SymbolGraph][CursorInfo] Add option to SourceKit's CursorInfo request to include the SymbolGraph JSON

Adds a new 'key.retrieve_symbol_graph' option to the request. When set to 1 it
includes the JSON for a SymbolGraph containing a single node for the symbol at
the requested position.

This also extends the SymbolGraph library with a new entry point to get a graph
for a single symbol, and to additionally support type substitution to match the
existing CursorInfo behavior (e.g. so that when invoked on `first` in
`Array<Int>().first`, the type is given as `Int?` rather than `Element?`).

Resolves rdar://problem/70551509

nathawes commented:
@swift-ci Please Build Toolchain macOS Platform

nathawes commented:
@swift-ci Please Build Toolchain macOS Platform

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci Please Build Toolchain macOS Platform

nathawes commented on lib/SymbolGraphGen/SymbolGraph.cpp:
```diff
@@ -499,7 +502,23 @@ void SymbolGraph::serialize(llvm::json::OStream &OS) {
 
     OS.attributeObject("module", [&](){
-      OS.attribute("name", M.getNameStr());
+      StringRef ModuleName = M.getNameStr();
+      SmallVector<Identifier, 4> Bystanders;
+
+      // If the decl is from a cross-import overlay module, report the overlay's
+      // declaring module as the owning module.
+      if (auto *Declaring = M.getDeclaringModuleIfCrossImportOverlay()) {
+        ModuleName = Declaring->getNameStr();
+        M.getRequiredBystandersIfCrossImportOverlay(Declaring, Bystanders);
+      }
+
+      OS.attribute("name", ModuleName);
+      if (!Bystanders.empty()) {
+        OS.attributeArray("bystanders", [&](){
+          for (auto &Id: Bystanders)
+            OS.value(Id.str());
+        });
+      }
```
This part will go away once https://github.com/apple/swift/pull/34922 lands.

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci Please Build Toolchain macOS Platform

Commit:
[test] Add missing requires line in test/SourceKit/CursorInfo/cursor_symbol_graph_objc.swift

It requires objc interop.

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci please test OS X Platform

nathawes commented:
@swift-ci please test Linux Platform

QuietMisdreavus reviewed (approved):
Symbol graph stuff looks good! One little nit in the new entry point, otherwise i think this is nice.

Commit:
[SymbolGraphGen] Address review comments on new printSymbolGraphForDecl entrypoint.

Change it to use EXIT_SUCCESS/FAILURE rather than a bool to match
emitSymbolGraphForModule.

nathawes commented:
@swift-ci please test

nathawes commented:
@akyrtzi I'm going to go ahead and merge this. If you have time to take a look I'll address any comments you have in a follow-up PR.

nathawes commented:
@swift-ci Please Build Toolchain macOS Platform
